app/src/main/java/org/imozerov/streetartview/ui/add/PickImageActivity.kt

```diff
@@ -48,18 +46,62 @@ class PickImageActivity : AppCompatActivity(), OnImagePickListener {
     override fun onOptionsItemSelected(item: MenuItem?): Boolean {
         if (item?.itemId == R.id.action_proceed) {
             startActivity(Intent(this, AddInfoActivity::class.java))
+        } else if (item?.itemId == R.id.action_camera) {
+            displayImages("Camera")
+        } else if (item?.itemId == R.id.action_pictures) {
+            displayImages("Pictures")
         }
         return super.onOptionsItemSelected(item)
     }
 
-    override fun onImagePicked(imageId: Int?) {
-        // TODO this needs to be full image but not just thumbnail
-        val uri = Uri.withAppendedPath(MediaStore.Images.Thumbnails.EXTERNAL_CONTENT_URI, imageId.toString())
-        image_picked.setImageURI(uri)
+    val IMAGES_ID = MediaStore.Images.Media._ID
+    val DISPLAY_NAME = MediaStore.Images.Media.DISPLAY_NAME
+    val BUCKET_ID = MediaStore.Images.Media.BUCKET_ID
+    val BUCKET_DISPLAY_NAME = MediaStore.Images.Media.BUCKET_DISPLAY_NAME
+    val DATE_TAKEN = MediaStore.Images.Media.DATE_TAKEN
+
+    private fun displayImages(bucket: String?) {
+        val projection = arrayOf<String>(IMAGES_ID,
+                DISPLAY_NAME,
+                BUCKET_ID,
+                BUCKET_DISPLAY_NAME,
+                DATE_TAKEN)
+
+        val cursor = contentResolver.query(
+                IMAGES_PATH,
+                projection,
+                null,
+                null,
+                DATE_TAKEN + " DESC")
+
+        adapter.updateDataset(cursorToList(cursor, bucket))
+    }
+
+    private fun cursorToList(cursor: Cursor, bucket: String?): List<PictureItem> {
+        val items = ArrayList<PictureItem>()
+
+        if (cursor.moveToFirst()) {
+            do {
+                val bucketName = cursor.getString(cursor.getColumnIndex(BUCKET_DISPLAY_NAME))
+                val id = cursor.getInt(cursor.getColumnIndex(IMAGES_ID))
+                if (bucket.equals(bucketName)) {
+                    items.add(PictureItem(id))
+                }
+            } while (cursor.moveToNext())
```

Comment:
Is it ok performance-wise to collect all data from cursor instead of working with cursor directly? 
Have you tested this solution on device with many images?